src/txHashBuilder.c

```diff
@@ -54,6 +54,138 @@ static void blake2b_256_append_cbor_tx_body(
 
 /* End of hash computation utilities. */
 
+static void cbor_append_txInput(tx_hash_builder_t *builder, const uint8_t *utxoHashBuffer, size_t utxoHashSize,
+                                uint32_t utxoIndex)
+{
+	// Array(2)[
+	//    Bytes[hash],
+	//    Unsigned[index]
+	// ]
+	BUILDER_APPEND_CBOR(CBOR_TYPE_ARRAY, 2);
+	{
+		ASSERT(utxoHashSize == TX_HASH_LENGTH);
+		BUILDER_APPEND_CBOR(CBOR_TYPE_BYTES, utxoHashSize);
+		BUILDER_APPEND_DATA(utxoHashBuffer, utxoHashSize);
+	}
+	{
+		BUILDER_APPEND_CBOR(CBOR_TYPE_UNSIGNED, utxoIndex);
+	}
+}
+
+static void cbor_append_legacy_txOutput(tx_hash_builder_t *builder, tx_hash_builder_output const *output)
+{
+	ASSERT(output->format == LEGACY);
+
+	if (output->numAssetGroups == 0) {
+		// Array(2 + includeDatumHash)[
+		//   Bytes[address]
+		//   Unsigned[amount]
+		//   //// entries added later, datum hash
+		// ]
+		BUILDER_APPEND_CBOR(CBOR_TYPE_ARRAY, 2 + output->includeDatumOption);
+		{
+			BUILDER_APPEND_CBOR(CBOR_TYPE_BYTES, output->addressSize);
+			BUILDER_APPEND_DATA(output->addressBuffer, output->addressSize);
+		}
+		{
+			BUILDER_APPEND_CBOR(CBOR_TYPE_UNSIGNED, output->amount);
+		}
+		builder->state = TX_HASH_BUILDER_IN_OUTPUTS_TOP_LEVEL_DATA;
+	} else {
+		builder->multiassetData.remainingAssetGroups = output->numAssetGroups;
+		// Array(2)[
+		//   Bytes[address]
+		//   Array(2)[]
```

Comment:
It seems this does not allow to add datum hash when tokens are present. That would be incorrect.